presto-main/src/main/java/com/facebook/presto/operator/aggregation/reservoirsample/ReservoirSamplingAggregationFunction.java

```diff
@@ -0,0 +1,267 @@
+        Class<? extends AccumulatorState> stateInterface;
+        AccumulatorStateSerializer<?> stateSerializer;
+
+        stateInterface = ReservoirSampleState.class;
+        stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+        inputFunction = LONG_INPUT_FUNCTION;
+        combineFunction = COMBINE_FUNCTION;
+        outputFunction = OUTPUT_FUNCTION;
+        /*
+        if (type.getJavaType() == long.class) {
+            stateInterface = ReservoirSampleState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = LONG_INPUT_FUNCTION;
+            combineFunction = COMBINE_FUNCTION;
+            outputFunction = OUTPUT_FUNCTION;
+        }
+        else if (type.getJavaType() == double.class) {
+            stateInterface = ReservoirSampleState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = DOUBLE_INPUT_FUNCTION;
+            combineFunction = DOUBLE_COMBINE_FUNCTION;
+            outputFunction = DOUBLE_OUTPUT_FUNCTION;
+        }
+        else if (type.getJavaType() == boolean.class) {
+            stateInterface = ReservoirSampleState.class;
+            stateSerializer = StateCompiler.generateStateSerializer(stateInterface, classLoader);
+            inputFunction = BOOLEAN_INPUT_FUNCTION;
+            combineFunction = BOOLEAN_COMBINE_FUNCTION;
+            outputFunction = BOOLEAN_OUTPUT_FUNCTION;
+        }
+        else {
+            // native container type is Slice or Block
+            stateInterface = BlockPositionState.class;
+            stateSerializer = new BlockPositionStateSerializer(type);
+            inputFunction = min ? BLOCK_POSITION_MIN_INPUT_FUNCTION.bindTo(type) : BLOCK_POSITION_MAX_INPUT_FUNCTION.bindTo(type);
+            combineFunction = min ? BLOCK_POSITION_MIN_COMBINE_FUNCTION.bindTo(type) : BLOCK_POSITION_MAX_COMBINE_FUNCTION.bindTo(type);
+            outputFunction = BLOCK_POSITION_OUTPUT_FUNCTION.bindTo(type);
+        }
+         */
+
+        AccumulatorStateFactory<?> stateFactory = StateCompiler.generateStateFactory(stateInterface, classLoader);
+
+        Type intermediateType = stateSerializer.getSerializedType();
+        final List<ParameterMetadata> parametersMetadata = createParameterMetadata(type);
+        AggregationMetadata metadata = new AggregationMetadata(
+                generateAggregationName(getSignature().getName(), type.getTypeSignature(), inputTypes.stream().map(Type::getTypeSignature).collect(toImmutableList())),
+                parametersMetadata,
+                inputFunction,
+                combineFunction,
+                outputFunction,
+                ImmutableList.of(new AccumulatorStateDescriptor(
+                        stateInterface,
+                        stateSerializer,
+                        stateFactory)),
+                type);
```

Comment:
@atavory 

> @mbasmanova I'd appreciate some help on writing an array output. In ReservoirSamplingAggregationFunction.java, 238, the call to BlockBuilder.beginBlockEntry fails, with an UnsupportedOperationException. Not sure why that is the case, nor what is the correct way to "open an array", followed by writing things into it in a loop, followed by "closing it".
> 
> Many thanks!

The problem is that the aggregation's declared return type is wrong. It is `type`, but needs to be `new Array(type)` on L172.

Reply:
Many thanks! Just to be sure, you meant `new ArrayType(type)`, right?

Reply:
@atavory 

> Just to be sure, you meant new ArrayType(type), right?

That's right.